### PR TITLE
Adding passive mode for DevPortal

### PR DESCRIPTION
**Pull request by Senthuran100**
Git issue : https://github.com/wso2/product-apim/issues/4171

### CODE REVIEW

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/services/login/login_callback.jag [hunk 1]
```diff
@@ -25,15 +25,21 @@
...
     }
-    if(request.getParameter("error") === "login_required"){
-       response.sendRedirect(serverUrl + app.context + "/services/configs");
-    } else if(request.getParameter("code") != null){
+    if(request.getParameter("error")==="login_required" && isPassive){
+        if(utils.getIdentityProviderConfig()!=null){
+        response.sendRedirect(serverUrl + app.context);
+        }
+        else{
+        response.content={error:"Loginerror"};
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/services/login/login_callback.jag]: Formatting issue.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/services/login/login_callback.jag [hunk 2]
```diff
@@ -25,15 +25,21 @@
...
         serverUrl = MGT_TRANSPORT + forwarded_for;
     } else {
         serverUrl = utils.getServerURL();
     }
-    if(request.getParameter("error") === "login_required"){
-       response.sendRedirect(serverUrl + app.context + "/services/configs");
-    } else if(request.getParameter("code") != null){
+    if(request.getParameter("error")==="login_required" && isPassive){
+        if(utils.getIdentityProviderConfig()!=null){
+        response.sendRedirect(serverUrl + app.context);
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/services/login/login_callback.jag]: Formatting issue.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/DevPortal.jsx [hunk 3]
```diff
@@ -84,6 +89,14 @@ class DevPortal extends React.Component {
         promisedSettings
             .then((response) => {
                 this.setSettings(response.body);
+            })
+            .then((response) => {
+                if(!this.state.settings.IsAnonymousModeEnabled){
+                    this.setState({isNonAnonymous:true})
+                }
+                if (Settings.app.isPassive && !AuthManager.getUser() && !sessionStorage.getItem('notEnoughPermission') && !this.state.isNonAnonymous) {
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/DevPortal.jsx]: Fix formatting. Exceeds 120 character limit.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/DevPortal.jsx [hunk 4]
```diff
@@ -244,14 +255,45 @@ class DevPortal extends React.Component {
         return (prefix + sufix);
     }
 
+    checkLoginUser(isExternalIDP) {
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/DevPortal.jsx]: Add method comments.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 5]
```diff
@@ -5780,6 +5780,29 @@ public String getRequestedTenant() {
...
+        try {
+            org.json.simple.JSONObject tenantConfig = APIUtil.getTenantConfig(tenantDomain);
+            Object value = tenantConfig.get(APIConstants.API_TENANT_CONF_ENABLE_ANONYMOUS_MODE);
+            if (value != null) {
+                return Boolean.parseBoolean(value.toString());
+            } else {
+                return APIUtil.isDevPortalAnonymous();
+            }
+        } catch (APIManagementException e) {
+            log.debug("Error while retrieving Recommendation config from registry", e);
```
Comment (menakaj) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: Why use debug logs?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java [hunk 6]
```diff
@@ -688,6 +689,7 @@ private Permissions() {
     public static final String API_STORE_MAP_EXISTING_AUTH_APPS = API_STORE + "MapExistingAuthApps";
     public static final String API_STORE_API_KEY_ALIAS = API_STORE + "ApiKeyAlias";
     public static final String WSO2_ANONYMOUS_USER = "wso2.anonymous.user";
+    public static final String API_DEVPORTAL_ANONYMOUS_MODE = API_STORE+"EnableAnonymousMode";
```
Comment (menakaj) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java]: Add space before and after '+'

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java [hunk 7]
```diff
@@ -10321,6 +10321,17 @@ public static boolean isCertificateExistsInTrustStore(X509Certificate certificat
...
+        APIManagerConfiguration config = ServiceReferenceHolder.getInstance().
+                getAPIManagerConfigurationService().getAPIManagerConfiguration();
+        String anonymousMode = config.getFirstProperty(APIConstants.API_DEVPORTAL_ANONYMOUS_MODE);
+        if (anonymousMode == null) {
+            return true;
+        }
+        return Boolean.parseBoolean(anonymousMode);
+    }
+
+
```
Comment (menakaj) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java]: Remove additional line

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/ProtectedApp.jsx [hunk 8]
```diff
@@ -21,9 +21,10 @@ import qs from 'qs';
 import { withTheme } from '@material-ui/core/styles';
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/ProtectedApp.jsx]: Please fix the following eslint issues.

/home/travis/build/wso2/carbon-apimgt/features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/ProtectedApp.jsx
   71:77  error  A space is required before '}'       object-curly-spacing
  180:1   error  More than 2 blank lines not allowed  no-multiple-empty-lines
  222:42  error  Operator '+' must be spaced          space-infix-ops

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/data/Constants.jsx [hunk 9]
```diff
@@ -20,6 +20,8 @@ const CONSTS = {
         INSUFFICIENT_PREVILEGES: '900403: Insufficient privileges to login',
         INVALID_TOKEN: '900401: Invalid token',
     },
+    ISLOGINPERMITTED:'isLoginPermitted',
+    LOGINSTATUS:'loginStatus',
```
Comment (malinthaprasan) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/data/Constants.jsx]: Please add spaces after ":"
    ISLOGINPERMITTED: 'isLoginPermitted',
    LOGINSTATUS: 'loginStatus',

**Comment by malinthaprasan** [comment]
@Senthuran100 can you kindly fix the conflicts?

**Comment by malinthaprasan** [comment]
@bhathiya kindly keep this open for today and @Senthuran100 will fix the conflicts.

**Comment by Arshardh** [comment]
Ack

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/ProtectedApp.jsx [hunk 10]
```diff
@@ -217,12 +215,12 @@ class ProtectedApp extends Component {
      */
     render() {
         const {
-            userResolved, tenantList, notEnoughPermission, tenantResolved, clientId
+            userResolved, tenantList, notEnoughPermission, tenantResolved, clientId,
         } = this.state;
         const checkSessionURL = Settings.idp.checkSessionEndpoint + '?client_id='
             + clientId + '&redirect_uri='+ window.location.origin
```
Comment (malinthaprasan) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/ProtectedApp.jsx:221]:
```suggestion
            + clientId + '&redirect_uri=' + window.location.origin
```

**Comment by tmkasun** [comment]
@malinthaprasan Shall we merge this PR ?

There is a conflict in the logic between this changes and https://github.com/wso2/carbon-apimgt/pull/8415 PR, I think we could hold merging this https://github.com/wso2/carbon-apimgt/pull/8415 until finding a proper fix WDYT? 

Here is the related issue for https://github.com/wso2/product-apim/issues/7874 getting an error when user try to view an API from devportal without login to it

**Comment by malinthaprasan** [comment]
> @malinthaprasan Shall we merge this PR ?
> 
> There is a conflict in the logic between this changes and #8415 PR, I think we could hold merging this #8415 until finding a proper fix WDYT?
> 
> Here is the related issue for [wso2/product-apim#7874](https://github.com/wso2/product-apim/issues/7874) getting an error when user try to view an API from devportal without login to it

+1 Was waiting till some eslint issues are fixed.. Let's merge now

**Comment by malinthaprasan** [comment]
> > @malinthaprasan Shall we merge this PR ?
> > There is a conflict in the logic between this changes and #8415 PR, I think we could hold merging this #8415 until finding a proper fix WDYT?
> > Here is the related issue for [wso2/product-apim#7874](https://github.com/wso2/product-apim/issues/7874) getting an error when user try to view an API from devportal without login to it
> 
> +1 Was waiting till some eslint issues are fixed.. Let's merge now

Please wait.. one more fix on the way by @Senthuran100 :)

**Pull request merged**